// @geocoder-free/utils/es/lib/HttpClient.js
import fetchNode from 'node-fetch';
import fetchJSON from 'fetch-jsonp';

function createHeader() {
  return Promise.resolve({
    Accept: 'application/json'
  });
}

function getFetchFunc(forceNodeFetch) {
  if (forceNodeFetch) {
    return fetchNode;
  }

  if (typeof window === 'undefined') {
    return fetchNode;
  }

  return fetchJSON;
}

async function Post(url, data, forceNodeFetch = false) {
  const headers = await createHeader();
  const fetch = getFetchFunc(forceNodeFetch);
  const res = await fetch(url, {
    credentials: 'same-origin',
    headers,
    method: 'POST',
    body: data
  });
  const result = await res.json();
  return result;
}

async function Get(url, returnType = 'json', forceNodeFetch = false) {
  const headers = await createHeader();
  let result;

  if (returnType === 'json') {
    // type=json returns fetch-jsonp
    const fetch = getFetchFunc(forceNodeFetch);
    const res = await fetch(url, {
      headers,
      method: 'GET'
    });
    result = res.json();
  } else {
    const res = await fetchNode(url, {
      headers,
      method: 'GET'
    });
    result = await res.text();
  }

  return result;
}

async function GetUrl(url) {
  const headers = await createHeader();
  const res = await fetchNode(url, {
    headers,
    method: 'GET'
  });
  return res.url;
}

export { Post, Get, GetUrl };
